lsif-semanticdb/src/main/java/com/sourcegraph/lsif_semanticdb/LsifSemanticdb.java

```diff
@@ -0,0 +1,132 @@
+    this.writer = writer;
+    this.options = options;
+    this.globals = new ConcurrentHashMap<>();
+  }
+
+  public static void run(LsifSemanticdbOptions options) throws IOException {
+    LsifWriter writer = new LsifWriter(options);
+    new LsifSemanticdb(writer, options).run();
+  }
+
+  private void run() throws IOException {
+    writer.emitMetaData();
+    int projectId = writer.emitProject(options.language);
+
+    List<Path> files = SemanticdbWalker.findSemanticdbFiles(options);
+    if (options.reporter.hasErrors()) return;
+    options.reporter.startProcessing(files.size());
+
+    Set<String> isExportedSymbol = exportSymbols(files);
+    List<Integer> documentIds =
+        filesStream(files)
+            .flatMap(d -> processPath(d, isExportedSymbol))
+            .collect(Collectors.toList());
+
+    writer.emitContains(projectId, documentIds);
+
+    writer.build();
+    options.reporter.endProcessing();
+  }
+
+  private Stream<Path> filesStream(List<Path> files) {
+    return options.parallel ? files.parallelStream() : files.stream();
+  }
+
+  private Set<String> exportSymbols(List<Path> files) {
+    return filesStream(files)
+        .flatMap(this::parseTextDocument)
+        .flatMap(
+            d ->
+                d.semanticdb.getOccurrencesList().stream()
+                    .filter(occ -> occ.getRole() == Role.DEFINITION)
+                    .map(SymbolOccurrence::getSymbol)
+                    .filter(SemanticdbSymbols::isGlobal))
+        .collect(Collectors.toSet());
+  }
+
+  private Stream<Integer> processPath(Path path, Set<String> isExportedSymbol) {
+    return parseTextDocument(path).map(d -> processDocument(d, isExportedSymbol));
+  }
+
+  private Integer processDocument(LsifTextDocument doc, Set<String> isExportedSymbol) {
+    int documentId = writer.emitDocument(doc);
+    Set<String> localDefinitions =
+        doc.semanticdb.getOccurrencesList().stream()
+            .filter(
+                occ ->
+                    occ.getRole() == Role.DEFINITION && SemanticdbSymbols.isLocal(occ.getSymbol()))
+            .map(SymbolOccurrence::getSymbol)
+            .collect(Collectors.toSet());
+    doc.id = documentId;
+    ResultSets results = new ResultSets(writer, globals, isExportedSymbol, localDefinitions);
+    Set<Integer> rangeIds = new LinkedHashSet<>();
+
+    for (SymbolOccurrence occ : doc.sortedSymbolOccurrences()) {
+      SymbolInformation symbolInformation =
+          doc.symbols.getOrDefault(occ.getSymbol(), SymbolInformation.getDefaultInstance());
+      ResultIds ids = results.getOrInsertResultSet(occ.getSymbol());
+      int rangeId = writer.emitRange(occ.getRange());
+      rangeIds.add(rangeId);
+
+      // Range
+      writer.emitNext(rangeId, ids.resultSet);
+
+      // Reference
+      writer.emitItem(ids.referenceResult, rangeId, doc.id);
+
+      // Definition
+      if (occ.getRole() == SymbolOccurrence.Role.DEFINITION && ids.isDefinitionDefined()) {
```

Comment:
in what case is `ids.isDefinitionDefined() == true` aka when would we have two identical definitions?

Reply:
The definition ID should always be available in this case, it's a bug if it's not. I updated the code to report an error message in this case instead of silently ignoring it.